source/extensions/filters/network/common/redis/fault_impl.cc

```diff
@@ -0,0 +1,136 @@
+  std::vector<std::string> commands;
+  for (const std::string& command : base_fault.commands()) {
+    commands.emplace_back(absl::AsciiStrToLower(command));
+  }
+  return commands;
+}
+
+FaultManagerImpl::FaultManagerImpl(
+    Random::RandomGenerator& random, Runtime::Loader& runtime,
+    const Protobuf::RepeatedPtrField<
+        ::envoy::extensions::filters::network::redis_proxy::v3::RedisProxy_RedisFault>
+        faults)
+    : fault_map_(buildFaultMap(faults)), random_(random), runtime_(runtime) {}
+
+FaultMap FaultManagerImpl::buildFaultMap(
+    const Protobuf::RepeatedPtrField<
+        ::envoy::extensions::filters::network::redis_proxy::v3::RedisProxy_RedisFault>
+        faults) {
+  // Next, create the fault map that maps commands to pointers to Fault objects.
+  // Group faults by command
+  FaultMap fault_map;
+  for (auto base_fault : faults) {
+    auto fault_ptr = std::make_shared<FaultImpl>(base_fault);
+    if (!fault_ptr->commands().empty()) {
+      for (const std::string& command : fault_ptr->commands()) {
+        fault_map[command].emplace_back(fault_ptr);
+      }
+    } else {
+      // Generic "ALL" entry in map for faults that map to all keys; also add to each command
+      fault_map[FaultManagerKeyNames::get().AllKey].emplace_back(fault_ptr);
+    }
+  }
+
+  // Add the ALL keys faults to each command too so that we can just query faults by command.
+  // Get all ALL_KEY faults.
+  FaultMap::iterator it_outer = fault_map.find(FaultManagerKeyNames::get().AllKey);
+  if (it_outer != fault_map.end()) {
+    for (const FaultSharedPtr& fault_ptr : it_outer->second) {
+      FaultMap::iterator it_inner;
+      for (it_inner = fault_map.begin(); it_inner != fault_map.end(); it_inner++) {
+        std::string command = it_inner->first;
+        if (command != FaultManagerKeyNames::get().AllKey) {
+          fault_map[command].push_back(fault_ptr);
+        }
+      }
+    }
+  }
+  return fault_map;
+}
+
+// Fault checking algorithm:
+//
+// For example, if we have an ERROR fault at 5% for all commands, and a DELAY fault at 10% for GET,
+// if we receive a GET, we want 5% of GETs to get DELAY, and 5% to get ERROR. Thus, we need to
+// amortize the percentages.
+//
+// 0. Get random number.
+// 1. Get faults for given command.s
+// 2. For each fault, calculate the amortized fault injection percentage.
+//
+// Note that we do not check to make sure the probabilities of faults are <= 100%!
+const Fault* FaultManagerImpl::getFaultForCommandInternal(std::string command) const {
+  FaultMap::const_iterator it_outer = fault_map_.find(command);
+  if (it_outer != fault_map_.end()) {
+    auto random_number = random_.random() % 100;
+    int amortized_fault = 0;
+
+    for (const FaultSharedPtr& fault_ptr : it_outer->second) {
+      uint64_t fault_injection_percentage =
+          runtime_.snapshot().getIntegerNumeratorOfFractionalPercent(
+              fault_ptr->runtimeKey().value(), fault_ptr->defaultValue());
```

Comment:
Can't you just use `featureEnabled` here and avoid adding a new API? This all still seems overly complicated to me. It seems like we should be able to just check each fault independently and avoid amortizing.

Reply:
Unfortunately no. If `featureEnabled` returns true for two different faults, I need to then choose which of those faults gets used somehow. Eventually, I need to do a non-boolean comparison.

In the case of 10% DELAY, 20% ERROR for a given command, if we a second probabilistic comparison after `featureEnabled`, that's no longer the same as saying each request has a 30% chance of a fault. Instead, we are saying 1.2 * 1.3 = 32% chance of a fault for each request.

Reply:
But why do we need to do that? This seems overly complicated per my other comments? We don't do anything like this in the HTTP fault filter.

Reply:
So as I understand it, the http fault filter checks first for the delay and then for the abort:

```
delay
delay -> abort
abort
```

If there are only two faults that works fine, but what we've done for Redis is different in that:
- We can have multiple error faults or delay faults, ie we could have 3+.
- Error faults have delay built in (in addition to a standalone delay fault that introduces no error)
- We plan to add two more faults - a cache miss, and connection failure.

If we had 3 faults (A, B, C), if we did a check like the http fault filter does, it gets complicated fast:

```
A
A -> B
A -> B -> C
A -> B
A -> C
B
B -> C
C
```

In this scheme we could certainly use `featureEnabled`, but we would need to walk each of the paths. The loop construction with amortization seems simpler to Henry and I.

Reply:
I think I'm missing something fundamental. Why can't you just do independent checks for each fault that is configured? I think that would be fine for most/all use cases?

Reply:
If we have 2 faults A, B with probabilities 10% and 50%, independent checks give us the wrong probabilities if we don't amortize. We expect 60% fault rate, but we get ~55%: `0.1 + 0.9*0.5 = 0.55`. For smaller probabilities of 10% or less, the differences are minor ie < 2% with two faults.

Reply:
As I already said, I don't think this complexity is really needed and it's pretty confusing, so I would rather just drop it. If you feel very strongly it is required, can you please figure out a way to do it without adding a new runtime API as I can't see any other use case for this API?

/wait

Reply:
I feel moderately strongly about it. I will move the function out of runtime and back into the fault code.